languages/python/sqlalchemy-oso/sqlalchemy_oso/roles.py

```diff
@@ -1,16 +1,142 @@
+                name = Column(String(256))
+
+    :param mutually_exclusive: Boolean flag that sets whether or not users \
+    can have more than one role for a given resource. Defaults to ``True``.
+    :type roles: bool
+
+    :return: the ResourceRole mixin, which must then be mixed into a SQLAlchemy model for the role. E.g.,
+
+        .. code-block:: python
+
+            OrganizationRoleMixin = oso_roles.resource_role_class(
+                Base, User, Organization, ["OWNER", "MEMBER", "BILLING"]
+            )
+
+            class OrganizationRole(Base, OrganizationRoleMixin):
+                pass
+
+
+    """
+    global ROLE_CLASSES
+    ROLE_CLASSES.append(
+        {
+            "user_model": user_model,
+            "resource_model": resource_model,
+        }
+    )
+
+    tablename = f"{resource_model.__name__.lower()}_roles"
+    if mutually_exclusive:
+        unique_constraint = UniqueConstraint(
+            f"{resource_model.__name__.lower()}_id", "user_id"
+        )
+    else:
+        unique_constraint = UniqueConstraint(
+            f"{resource_model.__name__.lower()}_id", "name", "user_id"
+        )
+
+    class ResourceRoleMixin:
+        choices = role_choices
+
+        __tablename__ = tablename
+        id = Column(Integer, primary_key=True)
+        name = Column(String())
+        __table_args__ = (unique_constraint,)
+
+        @validates("name")
+        def validate_name(self, key, name):
+            if name not in self.choices:
+                raise ValueError(
+                    f"{name} Is not a valid choice for {self.__class__.__name__}"
+                )
+            return name
+
+        @declared_attr
+        def user_id(self):
+            type = inspect(user_model).primary_key[0].type
+            name = inspect(user_model).primary_key[0].name
+            table_name = user_model.__tablename__
+            return Column(type, ForeignKey(f"{table_name}.{name}"))
+
+        @declared_attr
+        def user(self):
+            return relationship(user_model.__name__, backref=tablename)
+
+    @declared_attr
+    def resource_id(cls):
+        type = inspect(resource_model).primary_key[0].type
+        name = inspect(resource_model).primary_key[0].name
+        table_name = resource_model.__tablename__
+        return Column(type, ForeignKey(f"{table_name}.{name}"))
+
+    @declared_attr
+    def resource(cls):
+        return relationship(resource_model.__name__, backref="roles")
+
+    setattr(ResourceRoleMixin, f"{resource_model.__name__.lower()}_id", resource_id)
+    setattr(ResourceRoleMixin, resource_model.__name__.lower(), resource)
+
+    # Add the relationship between the user_model and the resource_model
+    resources = relationship(
+        resource_model.__name__,
+        secondary=tablename,
+        viewonly=True,
+        backref="users",
+        sync_backref=False,
+    )
+    # @Q: Do we try to pluralize this name correctly?
```

Comment:
Not for now, will add a task. Should document it somewhere though

Reply:
In the docstring for `resource_role_class`. Added a comment to the docs PR so we add it there too.